libtransact/src/workload/runner.rs

```diff
@@ -0,0 +1,425 @@
+    sender: Sender<ShutdownMessage>,
+}
+
+#[derive(Default)]
+struct WorkerBuilder {
+    id: Option<String>,
+    workload: Option<Box<dyn BatchWorkload>>,
+    targets: Option<Vec<String>>,
+    rate: Option<u64>,
+    auth: Option<String>,
+}
+
+impl WorkerBuilder {
+    /// Sets the ID of the worker
+    ///
+    /// # Arguments
+    ///
+    ///  * `id` - The unique ID of the worker
+    pub fn with_id(mut self, id: &str) -> WorkerBuilder {
+        self.id = Some(id.into());
+        self
+    }
+
+    /// Sets the workload that will be run against the targets
+    ///
+    /// # Arguments
+    ///
+    ///  * `workload` - The workload that will return batches to submit
+    pub fn with_workload(mut self, workload: Box<dyn BatchWorkload>) -> WorkerBuilder {
+        self.workload = Some(workload);
+        self
+    }
+
+    /// Sets the targets for the worker
+    ///
+    /// # Arguments
+    ///
+    ///  * `targets` - A list of URL for submitting the batches. The URL provided must be the full
+    ///               URL before adding `/batches` for submission
+    pub fn with_targets(mut self, targets: &[String]) -> WorkerBuilder {
+        self.targets = Some(targets.to_vec());
+        self
+    }
+
+    /// Sets the rate for the worker
+    ///
+    /// # Arguments
+    ///
+    ///  * `rate` - How many batches to submit per second
+    pub fn with_rate(mut self, rate: u64) -> WorkerBuilder {
+        self.rate = Some(rate);
+        self
+    }
+
+    /// Sets the auth for the worker
+    ///
+    /// # Arguments
+    ///
+    ///  * `auth` - The auth string to set against the Authorization header for the http request
+    pub fn with_auth(mut self, auth: &str) -> WorkerBuilder {
+        self.auth = Some(auth.to_string());
+        self
+    }
+
+    pub fn build(self) -> Result<Worker, WorkloadRunnerError> {
+        let id = self.id.ok_or_else(|| {
+            WorkloadRunnerError::WorkloadAddError(
+                "unable to build, missing field: `id`".to_string(),
+            )
+        })?;
+
+        let rate = self.rate.ok_or_else(|| {
+            WorkloadRunnerError::WorkloadAddError(
+                "unable to build, missing field: `rate`".to_string(),
+            )
+        })?;
+
+        let workload = self.workload.ok_or_else(|| {
+            WorkloadRunnerError::WorkloadAddError(
+                "unable to build, missing field: `workload`".to_string(),
+            )
+        })?;
+
+        let targets = self.targets.ok_or_else(|| {
+            WorkloadRunnerError::WorkloadAddError(
+                "unable to build, missing field: `target`".to_string(),
+            )
+        })?;
+
+        let auth = self.auth.ok_or_else(|| {
+            WorkloadRunnerError::WorkloadAddError(
+                "unable to build, missing field: `auth`".to_string(),
+            )
+        })?;
+
+        let (sender, receiver) = channel();
+
+        let time_to_wait = time::Duration::new(0, 1_000_000_000 / rate as u32);
```

Comment:
What is this `1_000_000_000` number? Should it be a `const`?

Reply:
updated to `let time_to_wait = time::Duration::from_secs(1) / rate;` for clarity